### PR TITLE
8276654: element-list order is non deterministic

### PR DESCRIPTION
**Pull request by andrew-m-leonard**
Fixes: https://bugs.openjdk.java.net/browse/JDK-8276654

A intermittent problem with the make dependencies means the jdk.javadoc element-list-.txt generation can remove the src defined element|package-list-<7,8,9,10>.txt files.
Recreatable by using --with-jobs=1 causing jdk.javadoc "gendata" to always occur after "java" module build dependency.

Signed-off-by: Andrew Leonard <anleonar@redhat.com>

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276654](https://bugs.openjdk.java.net/browse/JDK-8276654): element-list order is non deterministic

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6278/head:pull/6278` \
`$ git checkout pull/6278`

Update a local copy of the PR: \
`$ git checkout pull/6278` \
`$ git pull https://git.openjdk.java.net/jdk pull/6278/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6278`

View PR using the GUI difftool: \
`$ git pr show -t 6278`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6278.diff">https://git.openjdk.java.net/jdk/pull/6278.diff</a>

</details>

### CODE REVIEW

**Comment by erikj79** [comment]
Good catch, but wouldn't it be better if we could fix the recipe in jdk.javadoc/Gendata.gmk to only delete the files it intends to create. If I understand what files are created where correctly, you could generate a glob for the rm statement using something like this:

$(RM) $(ELEMENT_LISTS_DIR)/element-list-{$(call CommaList,$(call sequence,11,$(VERSION_FEATURE)))}.txt

**Comment by magicus** [comment]
@erikj79 Ah, we have a `sequence` function! I forgot about it. My initial reaction was also to prefer fixing the recipe to not delete non-related files, but I could not find a nice and not too overly complicated way of doing it. 

I agree that this is a better solution.

**Comment by andrew-m-leonard** [comment]
> @erikj79 Ah, we have a `sequence` function! I forgot about it. My initial reaction was also to prefer fixing the recipe to not delete non-related files, but I could not find a nice and not too overly complicated way of doing it.
> 
> I agree that this is a better solution.

@erikj79 Yes, I thought that to start with, but exactly as @magicus said, I couldn't figure and simple/non-complex way in gnu make language to fix it there either!

**Comment by andrew-m-leonard** [comment]
> Good catch, but wouldn't it be better if we could fix the recipe in jdk.javadoc/Gendata.gmk to only delete the files it intends to create. If I understand what files are created where correctly, you could generate a glob for the rm statement using something like this:
> 
> $(RM) $(ELEMENT_LISTS_DIR)/element-list-{$(call CommaList,$(call sequence,11,$(VERSION_FEATURE)))}.txt

I'll give that a try, cheers

**Comment by andrew-m-leonard** [comment]
@erikj79 yep looks good:
```
/usr/bin/rm -f /home/andrew/workspace/jdk/build/linux-x86_64-server-release/jdk/modules/jdk.javadoc/jdk/javadoc/internal/doclets/toolkit/resources/releases/element-list-{11,12,13,14,15,16,17,18}.txt
```

File: make/modules/jdk.javadoc/Gendata.gmk [hunk 1]
```diff
@@ -72,7 +72,7 @@ $(JDK_JAVADOC_DIR)/_element_lists.marker: \
     $(MODULE_INFOS)
 	$(call MakeTargetDir)
 	$(call LogInfo, Creating javadoc element lists)
-	$(RM) -r $(ELEMENT_LISTS_DIR)
+	$(RM) $(ELEMENT_LISTS_DIR)/element-list-{$(call CommaList,$(call sequence,$(GENERATE_SYMBOLS_FROM_JDK_VERSION),$(JDK_SOURCE_TARGET_VERSION)))}.txt
```
Comment (erikj79) [inline make/modules/jdk.javadoc/Gendata.gmk:75]: Good to see that it worked! I would only wish that you found a way to break up the line. Long lines make future side-by-side reviews and 3-way merges hard. We don't enforce strict 80, but try to stay in some reasonable ballpark in the build files.

I think both CommaList and sequence are ok with whitespace in their parameters. Otherwise you could pre-calculate the numbers list in a variable before the recipe.

Comment (andrew-m-leonard) [reply make/modules/jdk.javadoc/Gendata.gmk:75]: yep, i'll have a look to see what I can do, not helped with existing variables being so long!! (GENERATE_SYMBOLS_FROM_JDK_VERSION)

Comment (andrew-m-leonard) [reply make/modules/jdk.javadoc/Gendata.gmk]: yep, split on CommaList seems better and works fine, see revision

Comment (magicus) [reply make/modules/jdk.javadoc/Gendata.gmk]: Does it work if you have space after the comma in the call to sequence? We try to have a whitespace after comma where possible (and try to make function handle that case). If it does not work, we should really fix sequence...

**Review by jonathan-gibbons — COMMENTED** [review]
Can someone please verify that the contents of these files after the fix is the same as the contents of the files for "good" builds before the fix?

**Comment by andrew-m-leonard** [comment]
> Can someone please verify that the contents of these files after the fix is the same as the contents of the files for "good" builds before the fix?

@jonathan-gibbons done, all good

**Comment by magicus** [comment]
@andrew-m-leonard Please refrain from force-pushing to a PR. It makes the evolution of the patch hard to follow for reviewers.

I thought I had commented on the commas, but I can't find it now so maybe I failed to press the Comment button. So let me do it again.

Can you verify if it is possible to add whitespace after the commas in the "sequence" call? Normally, we use whitespace after all commas (except where impossible, like for built-in make functions). If that does not work, we need to fix the "sequence" method so it accepts whitespace after commas.

**Comment by andrew-m-leonard** [comment]
> @andrew-m-leonard Please refrain from force-pushing to a PR. It makes the evolution of the patch hard to follow for reviewers.
> 
> I thought I had commented on the commas, but I can't find it now so maybe I failed to press the Comment button. So let me do it again.
> 
> Can you verify if it is possible to add whitespace after the commas in the "sequence" call? Normally, we use whitespace after all commas (except where impossible, like for built-in make functions). If that does not work, we need to fix the "sequence" method so it accepts whitespace after commas.
> 
> Otherwise this looks good!

Yes splits fine, updated... thanks

File: make/modules/jdk.javadoc/Gendata.gmk [hunk 2]
```diff
@@ -72,7 +72,9 @@ $(JDK_JAVADOC_DIR)/_element_lists.marker: \
     $(MODULE_INFOS)
 	$(call MakeTargetDir)
 	$(call LogInfo, Creating javadoc element lists)
-	$(RM) -r $(ELEMENT_LISTS_DIR)
+	$(RM) $(ELEMENT_LISTS_DIR)/element-list-{$(call CommaList, \
+		$(call sequence,$(GENERATE_SYMBOLS_FROM_JDK_VERSION), \
```
Comment (magicus) [inline make/modules/jdk.javadoc/Gendata.gmk:76]:
```suggestion
		$(call sequence, $(GENERATE_SYMBOLS_FROM_JDK_VERSION), \
```
```suggestion
		$(call sequence,$(GENERATE_SYMBOLS_FROM_JDK_VERSION), \
```

Comment (andrew-m-leonard) [reply make/modules/jdk.javadoc/Gendata.gmk]: done

**Review by magicus — APPROVED** [review]
